### PR TITLE
ValueMapping: Added support for mapping text to color, boolean values, NaN and Null. Improved UI for value mapping.

### LINKED ISSUE #26837: Field Value Mapping Support for NaN values
**What would you like to be added**:
To be able to map the value "NaN" to some other text in the single state panel. So perhaps string insenstive "nan" for the value in a mapping corresponds to `isNaN`

**Why is this needed**:
To display NaN values as something else.

### LINKED ISSUE #33755: Mapping NaN
Hi,
I am calculating percentages via transformations.
I am getting NaN but unable to value map them .
Visualization - Stat
Value mapping has only two option - value/Range
number to text or text to number option is not coming.

Also for some calculation like 0/int Grafana is giving 0
for some 0/int it is giving NaN

NaN

### PR DESCRIPTION
**Pull request by torkelo**
Closes #33755
Closes #26837

* [x] Update definition model to include maps and ranges 
* [x] Migration to new model
* [x] Migration for value mappings in override rules 
* [x] New model no longer maps text to numbers (that are later processed with thresholds) 
* [x] Update view only editor (sidebar) UI 
* [x] New modal base edit table U
* [x] Add support for range mappings 
* [x] Use new color option in display processor 
* [x] New NullToText map type to map null, undefined an NaN 

WIP UI 

Screenshot from 2021-05-10 16-55-32

### CODE REVIEW

File: packages/grafana-data/src/field/displayProcessor.ts [hunk 1]
```diff
@@ -62,20 +64,28 @@ export function getDisplayProcessor(options?: DisplayProcessorOptions): DisplayP
...
+        if (mappingResult.state !== undefined) {
+          text = mappingResult.state;
+        }
 
-      if (mappedValue) {
-        text = mappedValue.text;
-        const v = isStringUnit ? NaN : toNumber(text);
+        if (mappingResult.value !== undefined && !isStringUnit) {
+          numeric = mappingResult.value;
+        }
```
Comment (ryantxu) [inline packages/grafana-data/src/field/displayProcessor.ts:85]: Is this really necessary?  We could remove the number option entirely now -- that can be supported via transformations if it is really required.  We always do `shouldFormat = false;` so the unit does not matter

**Comment by ryantxu** [comment]
@torkelo I think you are missing a file -- I don't see the compact view in history :man_shrugging: 

I changed the "NullMode" to "Special" and added true/false -- the dropdown needs some style help!  
image

File: packages/grafana-data/src/utils/anyToNumber.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  if (value === '' || value === null || value === undefined || Array.isArray(value)) {
+    return NaN; // lodash calls them 0
+  }
+
+  if (typeof value === 'boolean') {
+    return value ? 1 : 0;
+  }
+
+  return toNumber(value);
```
Comment (hugohaggmark) [inline packages/grafana-data/src/utils/anyToNumber.ts:17]: nit: Could we use Number instead?
```suggestion
  return Number(value);
```

File: packages/grafana-data/src/utils/valueMappings.ts [hunk 3]
```diff
@@ -1,104 +1,88 @@
...
-
-  if (isNumeric(value as string) && isNumeric(valueToTextMapping.value)) {
-    valueAsNumber = parseFloat(value as string);
-    valueToTextMappingAsNumber = parseFloat(valueToTextMapping.value as string);
-
-    if (valueAsNumber === valueToTextMappingAsNumber) {
-      return allValueMappings.concat(valueToTextMapping);
+import { ValueMapping, MappingType, ValueMappingResult, SpecialValueMatch } from '../types';
+
+export function getValueMappingResult(valueMappings: ValueMapping[], value: any): ValueMappingResult | null {
```
Comment (hugohaggmark) [inline packages/grafana-data/src/utils/valueMappings.ts:3]: Would a FieldRegistry for ValueMappings be a nicer approach?

File: public/app/plugins/panel/timeline/utils.ts [hunk 4]
```diff
@@ -74,8 +73,7 @@ export const preparePlotConfigBuilder: PrepConfig = ({
 
   const colorLookup = (seriesIdx: number, valueIdx: number, value: any) => {
     const field = frame.fields[seriesIdx];
-    const mode = field.config?.color?.mode;
-    if (mode && field.display && (mode === FieldColorModeId.Thresholds || mode.startsWith('continuous-'))) {
+    if (field.display) {
```
Comment (hugohaggmark) [inline public/app/plugins/panel/timeline/utils.ts:76]: Hmm, how does this change relate to value mapping?

**Review by hugohaggmark — APPROVED** [review]
Looks great, going to take it for a test spin after this!

File: packages/grafana-data/src/utils/anyToNumber.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-data/src/utils/anyToNumber.ts:17]: not sure, this was only copied code. Think there are some differences with toNumber vs Number constructor

File: packages/grafana-data/src/utils/valueMappings.ts (hunk 3 above)
Comment (torkelo) [reply packages/grafana-data/src/utils/valueMappings.ts:3]: maybe, but do no foresee many of these mapping types, think these 3 will suffice (famous last words).

**Review by hugohaggmark — COMMENTED** [review]
Tested it with Postgres with different scenarios and works as expected!

**Review by jackw — APPROVED** [review]
I'm not all that familiar with the feature but tested it locally with stat panel and null, range and value mappings and works as expected. 🚀

File: packages/grafana-ui/src/components/ValueMappingsEditor/ValueMappingsEditor.tsx [hunk 5]
```diff
@@ -1,60 +1,90 @@
...
+    </VerticalGroup>
   );
-};
+});
+
+ValueMappingsEditor.displayName = 'ValueMappingsEditor';
+
+export const getStyles = (theme: GrafanaTheme2) => ({
+  modal: css({
+    width: '980px',
```
[the remark this answers is not in the document]
Comment (torkelo) [reply packages/grafana-ui/src/components/ValueMappingsEditor/ValueMappingsEditor.tsx:82]: there is a max-width: 100% on the default modal styles that makes this work / adapt even to smaller screens

**Comment by leeoniya** [comment]
nice.

**Comment by oddlittlebird** [comment]
@torkelo, are you planning to write docs for this?

**Review by torkelo — COMMENTED** [review]
The docs should already be updated as part of this PR. But only bare minimum update. Think the docs is missing the special type, that was added after the docs update and missed that.

**Comment by oddlittlebird** [comment]
Ah, I see

**Pull request merged**